nomos-libp2p/src/behaviour/nat/address_mapper/protocols/pcp/connection.rs

```diff
@@ -0,0 +1,93 @@
+use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+
+use backon::{ExponentialBuilder, Retryable as _};
+use rand::RngCore as _;
+use tokio::{
+    net::UdpSocket,
+    time::{timeout, Duration},
+};
+use zerocopy::{Immutable, IntoBytes};
+
+use crate::behaviour::nat::address_mapper::protocols::pcp::client::{PcpConfig, PcpError};
+
+const BUFFER_SIZE: usize = 1024;
+/// Let OS choose any available port for client socket
+const ANY_PORT: u16 = 0;
+
+/// PCP connection for reliable communication with retry logic
+///
+/// Handles connection to PCP server with exponential backoff retry.
+/// Each connection is bound to client IP and connected to server.
+pub struct PcpConnection {
+    socket: UdpSocket,
+    config: PcpConfig,
+}
+
+impl PcpConnection {
+    pub async fn connect(
+        client_addr: Ipv4Addr,
+        gateway_ip: Ipv4Addr,
+        config: PcpConfig,
+    ) -> Result<Self, PcpError> {
+        let socket = UdpSocket::bind((client_addr, ANY_PORT)).await?;
+
+        let server_addr = SocketAddr::new(IpAddr::V4(gateway_ip), super::client::PCP_PORT);
+        socket.connect(server_addr).await?;
+
+        Ok(Self { socket, config })
+    }
+
+    pub async fn send_request<Request, Response>(
+        &self,
+        request: &Request,
+    ) -> Result<Response, PcpError>
+    where
+        Request: IntoBytes + Immutable + Sync,
+        Response: for<'a> TryFrom<&'a [u8], Error = PcpError>,
+    {
+        let backoff = create_backoff(
+            self.config.initial_retry,
+            self.config.max_retry_delay,
+            self.config.max_retries,
+        );
+
+        (|| async {
+            self.socket.send(request.as_bytes()).await?;
+
+            let mut buffer = vec![0u8; BUFFER_SIZE];
+
+            match timeout(self.config.request_timeout, self.socket.recv(&mut buffer)).await {
+                Ok(Ok(size)) => {
+                    buffer.truncate(size);
+                    Response::try_from(buffer.as_slice())
```

Comment:
Just a nit, I think we could also not require the response to implement `TryFrom` with a `PcpError`, but could map it ourselves internally into an `InvalidResponse` variant. But this is also ok, although somewhat slightly less flexible.